### PR TITLE
feat: ContributionResolver - createContribution

### PR DESCRIPTION
**Pull request by Elweyn**
## 🍰 Pullrequest
First call to create Contribution.

### Issues

- relates #2000

### Todo

- [X] Create call createContribution
- [X] Test createContribution

### CODE REVIEW

File: backend/src/graphql/resolver/AdminResolver.ts [hunk 1]
```diff
@@ -46,11 +46,17 @@ import { checkOptInCode, activationLink, printTimeDuration } from './UserResolve
...
 import { transactionLinkCode as contributionLinkCode } from './TransactionLinkResolver'
 import CONFIG from '@/config'
+import { getUserCreation } from './util/getUserCreation'
+import { isContributionValid } from './util/isContributionValid'
 
 // const EMAIL_OPT_IN_REGISTER = 1
 // const EMAIL_OPT_UNKNOWN = 3 // elopage?
 const MAX_CREATION_AMOUNT = new Decimal(1000)
-const FULL_CREATION_AVAILABLE = [MAX_CREATION_AMOUNT, MAX_CREATION_AMOUNT, MAX_CREATION_AMOUNT]
+export const FULL_CREATION_AVAILABLE = [
```
Comment (Mogge) [inline backend/src/graphql/resolver/AdminResolver.ts:55]: Can we have this in a `constants` folder?

File: backend/src/graphql/resolver/AdminResolver.ts [hunk 2]
```diff
@@ -689,13 +695,10 @@ interface CreationMap {
...
 }
 
-export const getUserCreation = async (id: number, includePending = true): Promise<Decimal[]> => {
-  logger.trace('getUserCreation', id, includePending)
-  const creations = await getUserCreations([id], includePending)
-  return creations[0] ? creations[0].creations : FULL_CREATION_AVAILABLE
-}
-
-async function getUserCreations(ids: number[], includePending = true): Promise<CreationMap[]> {
+export async function getUserCreations(
```
Comment (Mogge) [inline backend/src/graphql/resolver/AdminResolver.ts:698]: Why did you change this. We use the `=>` notations for functions

Comment (Mogge) [reply backend/src/graphql/resolver/AdminResolver.ts:698]: Move this to `utils` as well?

File: backend/src/graphql/resolver/ContributionResolver.test.ts [hunk 3]
```diff
@@ -0,0 +1,183 @@
...
+          }),
+        ).resolves.toEqual(
+          expect.objectContaining({
+            errors: [new GraphQLError('401 Unauthorized')],
+          }),
+        )
+      })
+    })
+
+    describe('authenticated with deleted user', () => {
```
Comment (Mogge) [inline backend/src/graphql/resolver/ContributionResolver.test.ts:49]: This is not needed. Done by the user resolver tests, as a deleted user cannot login

File: backend/src/graphql/resolver/ContributionResolver.test.ts [hunk 4]
```diff
@@ -0,0 +1,183 @@
...
+            },
+          }),
+        ).resolves.toEqual(
+          expect.objectContaining({
+            errors: [new GraphQLError('No information for available creations for the given date')],
+          }),
+        )
+      })
+
+      it('creates contribution', async () => {
```
Comment (Mogge) [inline backend/src/graphql/resolver/ContributionResolver.test.ts:163]: describe block `valid input`

File: backend/src/graphql/resolver/ContributionResolver.ts [hunk 5]
```diff
@@ -0,0 +1,35 @@
...
+    const contribution = Contribution.create()
+    contribution.userId = user.id
+    contribution.amount = amount
+    contribution.createdAt = new Date()
+    contribution.contributionDate = creationDateObj
+    contribution.memo = memo
+
+    logger.trace('contribution to save', contribution)
+    await Contribution.save(contribution)
+    return true
```
Comment (Mogge) [inline backend/src/graphql/resolver/ContributionResolver.ts:33]:
```suggestion
    return new UnconfirmedContribution(contribution)
```

File: backend/src/graphql/resolver/util/isContributionValid.ts [hunk 6]
```diff
@@ -0,0 +1,24 @@
+import { backendLogger as logger } from '@/server/logger'
+import Decimal from 'decimal.js-light'
+import { getCreationIndex } from '../AdminResolver'
+
+export const isContributionValid = (
+  creations: Decimal[],
+  amount: Decimal,
+  creationDate: Date,
+): boolean => {
```
Comment (Mogge) [inline backend/src/graphql/resolver/util/isContributionValid.ts:9]:
```suggestion
): void => {
```

File: backend/src/graphql/resolver/util/isContributionValid.ts [hunk 7]
```diff
@@ -0,0 +1,24 @@
...
+    throw new Error('No information for available creations for the given date')
+  }
+
+  if (amount.greaterThan(creations[index].toString())) {
+    throw new Error(
+      `The amount (${amount} GDD) to be created exceeds the amount (${creations[index]} GDD) still available for this month.`,
+    )
+  }
+
+  return true
```
[the remark this answers is not in the document]
Comment (Mogge) [reply backend/src/graphql/resolver/util/isContributionValid.ts:23]: This function throws an error, when the contribution is not valid. No return value needed, as it would always return true if it does not throw.

**Review by Mogge — CHANGES_REQUESTED** [review]

File: backend/src/graphql/resolver/AdminResolver.ts (hunk 2 above)
Comment (Elweyn) [reply backend/src/graphql/resolver/AdminResolver.ts:698]: Yes, was the refactoring function of my ide will, I will change this!

File: backend/src/graphql/resolver/AdminResolver.ts (hunk 1 above)
Comment (Elweyn) [reply backend/src/graphql/resolver/AdminResolver.ts:55]: Sure do we have one or should I set the const folder on same level as the util one?

File: backend/src/graphql/resolver/TransactionLinkResolver.ts [hunk 8]
```diff
@@ -28,7 +28,8 @@ import { executeTransaction } from './TransactionResolver'
 import { Order } from '@enum/Order'
 import { Contribution as DbContribution } from '@entity/Contribution'
 import { ContributionLink as DbContributionLink } from '@entity/ContributionLink'
-import { getUserCreation, isContributionValid } from './AdminResolver'
+import { isContributionValid } from './util/isContributionValid'
+import { getUserCreation } from './util/getUserCreation'
```
[the remark this answers is not in the document]
Comment (Elweyn) [reply backend/src/graphql/resolver/TransactionLinkResolver.ts:32]: OK, but how would you name the file then? util/validCreation oder validContribution?

File: backend/src/graphql/resolver/ContributionResolver.ts [hunk 9]
```diff
@@ -0,0 +1,35 @@
...
+    @Args() { amount, memo, creationDate }: CreateContributionArgs,
+    @Ctx() context: Context,
+  ): Promise<boolean> {
+    const user = getUser(context)
+    const creations = await getUserCreation(user.id)
+    logger.trace('creations', creations)
+    const creationDateObj = new Date(creationDate)
+    isContributionValid(creations, amount, creationDateObj)
+
+    const contribution = Contribution.create()
```
[the remark this answers is not in the document]
Comment (Elweyn) [reply backend/src/graphql/resolver/ContributionResolver.ts:24]: I'm not sure to know the difference! In my opinion the one creates an object in the db and makes it to an JS Object (Contribution.create) and the other only create an object of the db class (new Contribution).

File: backend/src/graphql/resolver/util/isContributionValid.ts [hunk 10]
```diff
@@ -0,0 +1,119 @@
...
+  })
+}
+
+export const getUserCreation = async (id: number, includePending = true): Promise<Decimal[]> => {
+  logger.trace('getUserCreation', id, includePending)
+  const creations = await getUserCreations([id], includePending)
+  return creations[0] ? creations[0].creations : FULL_CREATION_AVAILABLE
+}
+
+export const getCreationMonths = (): number[] => {
```
Comment (Mogge) [inline backend/src/graphql/resolver/util/isContributionValid.ts:87]: Why export?

File: backend/src/graphql/resolver/util/isContributionValid.ts [hunk 11]
```diff
@@ -0,0 +1,119 @@
...
+  const now = new Date(Date.now())
+  return [
+    now.getMonth() + 1,
+    new Date(now.getFullYear(), now.getMonth() - 1, 1).getMonth() + 1,
+    new Date(now.getFullYear(), now.getMonth() - 2, 1).getMonth() + 1,
+  ].reverse()
+}
+
+export const getCreationIndex = (month: number): number => {
```
Comment (Mogge) [inline backend/src/graphql/resolver/util/isContributionValid.ts:96]: Why export?

File: backend/src/graphql/resolver/util/isContributionValid.ts [hunk 12]
```diff
@@ -0,0 +1,119 @@
+import { TransactionTypeId } from '@/graphql/enum/TransactionTypeId'
```
Comment (Mogge) [inline backend/src/graphql/resolver/util/isContributionValid.ts:1]: Think of renaming this file.

**Review by Mogge — CHANGES_REQUESTED** [review]

**Review by Mogge — APPROVED** [review]
